File: rxpermission/src/main/java/com/orca/kam/rxpermission/commons/PermissionActivity.java
package com.orca.kam.rxpermission.commons;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Window;

import com.afollestad.materialdialogs.MaterialDialog;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import java.util.ArrayList;
import java.util.List;

import static android.content.pm.PackageManager.PERMISSION_DENIED;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;
import static android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS;
import static android.provider.Settings.ACTION_MANAGE_APPLICATIONS_SETTINGS;
import static android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE;
import static com.orca.kam.rxpermission.commons.PermissionContent.KEY_PERMISSION_CONTENT;

/**
 * Project RxPermission
 *
 * @author Kang Young Won
 * @create 2016-08-29 - 오후 12:43
 */

public class PermissionActivity extends AppCompatActivity {

    private static final String TAG = "PermissionActivity";

    private static final int REQ_CODE_PERMISSION_REQUEST = 10;
    private static final int REQ_CODE_REQUEST_SETTING = 20;

    private PermissionContent content;
    private List<String> permissions;


    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setWindowUntouchable();
        createPermissionContent();
        checkPermissions();
    }


    private void setWindowUntouchable() {
        Window window = getWindow();
        if (window != null) window.addFlags(FLAG_NOT_TOUCHABLE);
    }


    private void createPermissionContent() {
        content = getExtraBundle().getParcelable(KEY_PERMISSION_CONTENT);
        Preconditions.checkArgument(content != null, "Invalid PermissionContent");
        permissions = content.getPermissions();
    }


    private Bundle getExtraBundle() {
        Bundle bundle = getIntent().getExtras();
        Preconditions.checkArgument((bundle != null && !bundle.isEmpty()), "Invalid Bundle(Extras)");
        return bundle;
    }


    private void checkPermissions() {
        log("checkPermissions");
        List<String> deniedPermission = getDeniedPermissionList();
        if (deniedPermission.isEmpty()) {
            permissionGranted();
        } else {
            showRationaleDialog(deniedPermission);
        }
    }


    @Override public void onRequestPermissionsResult(
            int requestCode, @NonNull String[] permissionResults, @NonNull int[] grantResults) {
        log("onRequestPermissionsResult");
        if (requestCode == REQ_CODE_PERMISSION_REQUEST) {
            List<String> deniedPermissions = getDeniedPermissionList(permissionResults, grantResults);
            if (deniedPermissions.isEmpty()) {
                permissionGranted();
            } else {
                showPermissionDenyDialog(deniedPermissions);
            }
        }
    }


    @Override public void finish() {
        super.finish();
        overridePendingTransition(0, 0);
    }


    private void requestPermissions(List<String> needPermissions) {
        log("requestPermissions");
        ActivityCompat.requestPermissions(this,
                needPermissions.toArray(new String[needPermissions.size()]),
                REQ_CODE_PERMISSION_REQUEST);
    }


    private void showRationaleDialog(final List<String> needPermissions) {
        log("showRationaleDialog");
        new MaterialDialog.Builder(this)
                .content(content.getExplanationMessage())
                .negativeText(content.getExplanationConfirmButtonText())
                .onNegative((dialog, which) -> requestPermissions(needPermissions))
                .cancelable(false)
                .show();
    }


    private void showPermissionDenyDialog(final List<String> deniedPermissions) {
        log("showPermissionDenyDialog");
        new MaterialDialog.Builder(this)
                .content(content.getDeniedMessage())
                .positiveText(content.getSettingButtonText())
                .negativeText(content.getDeniedCloseButtonText())
                .onPositive((dialog, which) -> showAppSettings())
                .onNegative((dialog, which) -> permissionDenied(deniedPermissions))
                .cancelable(false)
                .show();
    }


    private void showAppSettings() {
        Intent intent = new Intent();
        String packageName = content.getPackageName();
        if (!Strings.isNullOrEmpty(packageName)) {
            intent.setAction(ACTION_APPLICATION_DETAILS_SETTINGS)
                    .setData(Uri.parse("package:" + content.getPackageName()));
        } else {
            intent.setAction(ACTION_MANAGE_APPLICATIONS_SETTINGS);
        }
        startActivityForResult(intent, REQ_CODE_REQUEST_SETTING);
    }


    private List<String> getDeniedPermissionList(String[] permissions, int[] grantResults) {
        List<String> deniedPermissions = new ArrayList<>();
        for (int i = 0; i < permissions.length; i++) {
            if (grantResults[i] == PERMISSION_DENIED) {
                deniedPermissions.add(permissions[i]);
            }
        }
        return deniedPermissions;
    }


    private List<String> getDeniedPermissionList() {
        return Lists.newArrayList(Iterables.filter(permissions, permission -> !isGrantedPermission(permission)));
    }


    private boolean isGrantedPermission(String permission) {
        return ContextCompat.checkSelfPermission(this, permission) == PERMISSION_GRANTED;
    }


    private void permissionGranted() {
        log("permissionGranted");
        AndroidPermission.permissionGranted();
        finish();
    }


    private void permissionDenied(List<String> deniedPermissions) {
        log("permissionDenied");
        AndroidPermission.permissionDenied(deniedPermissions);
        finish();
    }


    private void log(String msg) {
        Log.e(TAG, msg);
    }
}

File: rxpermissionkt/src/main/java/com/orca/kam/rxpermissionkt/commons/PermissionActivity.kt
package com.orca.kam.rxpermissionkt.commons

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.view.Window

import com.afollestad.materialdialogs.MaterialDialog
import com.f2prateek.dart.Dart
import com.f2prateek.dart.InjectExtra
import com.google.common.base.Strings
import com.google.common.collect.Iterables
import com.google.common.collect.Lists

import java.util.ArrayList

import android.content.pm.PackageManager.PERMISSION_DENIED
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS
import android.provider.Settings.ACTION_MANAGE_APPLICATIONS_SETTINGS
import android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE

/**
 * Project RxPermission

 * @author Kang Young Won
 * *
 * @create 2016-08-29 - 오후 12:43
 */

class PermissionActivity : AppCompatActivity() {

    @InjectExtra internal var content: PermissionContent? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Dart.inject(this)
        setWindowUntouchable()
        checkPermissions()
    }


    private fun setWindowUntouchable() {
        val window = window
        window?.addFlags(FLAG_NOT_TOUCHABLE)
    }


    private fun checkPermissions() {
        val deniedPermission = getDeniedPermissionList(content!!.permissions)
        if (deniedPermission.isEmpty()) {
            permissionGranted()
        } else {
            showRationaleDialog(deniedPermission)
        }
    }


    override fun onRequestPermissionsResult(
            requestCode: Int, permissionResults: Array<String>, grantResults: IntArray) {
        if (requestCode == REQ_CODE_PERMISSION_REQUEST) {
            val deniedPermissions = getDeniedPermissionList(permissionResults, grantResults)
            if (deniedPermissions.isEmpty()) {
                permissionGranted()
            } else {
                showPermissionDenyDialog(deniedPermissions)
            }
        }
    }


    override fun finish() {
        super.finish()
        overridePendingTransition(0, 0)
    }


    private fun requestPermissions(needPermissions: List<String>) {
        ActivityCompat.requestPermissions(this,
                needPermissions.toTypedArray(),
                REQ_CODE_PERMISSION_REQUEST)
    }


    private fun showRationaleDialog(needPermissions: List<String>) {
        MaterialDialog.Builder(this)
                .content(content!!.getExplanationMessage())
                .negativeText(content!!.getExplanationConfirmButtonText())
                .onNegative { dialog, which -> requestPermissions(needPermissions) }
                .cancelable(false)
                .show()
    }


    private fun showPermissionDenyDialog(deniedPermissions: List<String>) {
        MaterialDialog.Builder(this)
                .content(content!!.getDeniedMessage())
                .positiveText(content!!.getSettingButtonText())
                .negativeText(content!!.getDeniedCloseButtonText())
                .onPositive { dialog, which -> showAppSettings() }
                .onNegative { dialog, which -> permissionDenied(deniedPermissions) }
                .cancelable(false)
                .show()
    }


    private fun showAppSettings() {
        val intent = Intent()
        val packageName = content!!.packageName
        if (!Strings.isNullOrEmpty(packageName)) {
            intent.setAction(ACTION_APPLICATION_DETAILS_SETTINGS).data = Uri.parse("package:" + content!!.packageName!!)
        } else {
            intent.action = ACTION_MANAGE_APPLICATIONS_SETTINGS
        }
        startActivityForResult(intent, REQ_CODE_REQUEST_SETTING)
    }


    private fun getDeniedPermissionList(permissions: Array<String>, grantResults: IntArray): List<String> {
        val deniedPermissions = ArrayList<String>()
        for (i in permissions.indices) {
            if (grantResults[i] == PERMISSION_DENIED) {
                deniedPermissions.add(permissions[i])
            }
        }
        return deniedPermissions
    }


    private fun getDeniedPermissionList(permissions: List<String>): List<String> {
        return Lists.newArrayList(Iterables.filter(permissions) { permission -> !isGrantedPermission(permission) })
    }


    private fun isGrantedPermission(permission: String): Boolean {
        return ContextCompat.checkSelfPermission(this, permission) == PERMISSION_GRANTED
    }


    private fun permissionGranted() {
        AndroidPermission.permissionGranted()
        finish()
    }


    private fun permissionDenied(deniedPermissions: List<String>) {
        AndroidPermission.permissionDenied(deniedPermissions)
        finish()
    }

    companion object {


        private val REQ_CODE_PERMISSION_REQUEST = 10
        private val REQ_CODE_REQUEST_SETTING = 20
    }
}
